### PR TITLE
Kernel memory configs

### PR DESCRIPTION
**Pull request by yuumasato**
#### Description:

Adds rules to check for the following  kernel build configs:
- CONFIG_ARCH_HAS_STRICT_KERNEL_RWX=y
- CONFIG_DEBUG_WX=y
- CONFIG_DEBUG_FS=n
- CONFIG_STACKPROTECTOR=y
- CONFIG_STACKPROTECTOR_STRONG=y
- CONFIG_SCHED_STACK_END_CHECK=y
- CONFIG_HARDENED_USERCOPY=y
- #CONFIG_HARDENED_USERCOPY_FALLBACK is not set
- CONFIG_VMAP_STACK=y
- CONFIG_REFCOUNT_FULL=y
- CONFIG_FORTIFY_SOURCE = y
- #CONFIG_ACPI_CUSTOM_METHOD is not set
- #CONFIG_DEVKMEM is not set
- #CONFIG_PROC_KCORE is not set
- #CONFIG_COMPAT_VDSO is not set
- CONFIG_SECURITY_DMESG_RESTRICT=y
- CONFIG_RETPOLINE=y
- CONFIG_LEGACY_VSYSCALL_NONE=y
- CONFIG_LEGACY_VSYSCALL_EMULATE=n
- CONFIG_LEGACY_VSYSCALL_XONLY=n
- CONFIG_X86_VSYSCALL_EMULATION=n

### CODE REVIEW

File: linux_os/guide/system/kernel_build_config/kernel_config_fortify_source/rule.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+documentation_complete: true
+
+prodtype: fedora,rhel8,rhel9
+
+title: 'Harden common str/mem functions agains buffer overflows'
```
Comment (jan-cerny) [inline linux_os/guide/system/kernel_build_config/kernel_config_fortify_source/rule.yml:5]:
```suggestion
title: 'Harden common str/mem functions against buffer overflows'
```

File: linux_os/guide/system/kernel_build_config/kernel_config_legacy_vsyscall_xonly/rule.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+documentation_complete: true
+
+prodtype: fedora,rhel9
+
+title: 'Disable vsyscall emluate execution only'
```
Comment (jan-cerny) [inline linux_os/guide/system/kernel_build_config/kernel_config_legacy_vsyscall_xonly/rule.yml:5]:
```suggestion
title: 'Disable vsyscall emulate execution only'
```

File: linux_os/guide/system/kernel_build_config/kernel_config_refcount_full/rule.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
+documentation_complete: true
+
+prodtype: fedora,rhel7,rhel8,rhel9
+
+title: 'Perform full reference count validation'
+
+description: |-
+    Enabling this switches the refcounting infrastructure from a fast unchecked atomic_t
+    implementation to a fully state checked implementation, which can be have a slight
```
Comment (jan-cerny) [inline linux_os/guide/system/kernel_build_config/kernel_config_refcount_full/rule.yml:9]:
```suggestion
    implementation to a fully state checked implementation, which can have a slight
```

File: linux_os/guide/system/kernel_build_config/kernel_config_retpoline/rule.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+description: |-
+    Compile kernel with the retpoline compiler options to guard against kernel-to-user data leaks
+    by avoiding speculative indirect branches.
+    Requires a compiler with -mindirect-branch=thunk-extern support for full protection.
+    The kernel may run slower.
+
+    {{{ kernel_build_config_describe_config("CONFIG_RETPOLINE", "y") | indent(4) }}}
+
+rationale: |-
+    This is required to enable protection against Specter v2.
```
Comment (jan-cerny) [inline linux_os/guide/system/kernel_build_config/kernel_config_retpoline/rule.yml:14]:
```suggestion
    This is required to enable protection against Spectre v2.
```

File: linux_os/guide/system/kernel_build_config/kernel_config_sched_stack_end_check/rule.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
+documentation_complete: true
+
+prodtype: fedora,rhel8,rhel9
+
+title: 'Detect stack corruption on calls to schedule()'
+
+description: |-
+    This option checks for a stack overrun on calls to schedule(). If the stack end location is
+    found to be over written always panic as the content of the corrupted region can no longer
```
Comment (jan-cerny) [inline linux_os/guide/system/kernel_build_config/kernel_config_sched_stack_end_check/rule.yml:9]:
```suggestion
    found to be overwritten always panic as the content of the corrupted region can no longer
```

File: linux_os/guide/system/kernel_build_config/kernel_config_security_dmesg_restrict/rule.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+title: 'Restrict unprivileged access to the kernel syslog'
+
+description: |-
+    Enforce restrictions on unprivileged users reading the kernel syslog via dmesg(8).
+
+    {{{ kernel_build_config_describe_config("CONFIG_SECURITY_DMESG_RESTRICT", "n") | indent(4) }}}
+
+rationale: |-
+    Prevents unprivileted users from retrieving kernel addresses with dmesg.
```
Comment (jan-cerny) [inline linux_os/guide/system/kernel_build_config/kernel_config_security_dmesg_restrict/rule.yml:11]:
```suggestion
    Prevents unprivileged users from retrieving kernel addresses with dmesg.
```

**Review by jan-cerny — CHANGES_REQUESTED** [review]

File: linux_os/guide/system/kernel_build_config/kernel_config_compat_vdso/rule.yml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+documentation_complete: true
+
+title: 'Disable the 32-bit vDSO'
+
+description: |-
+    Map the 32-bit VDSO to the predictable old-style address too.
```
[the remark this answers is not in the document]
Comment (yuumasato) [reply linux_os/guide/system/kernel_build_config/kernel_config_compat_vdso/rule.yml]: I have extended the description.

**Review by jan-cerny — APPROVED** [review]
The fail of test scenarios is expected because these rules aren't testable in a container environment but they all use a template that has already been tested.

**Review by ggbecker — COMMENTED** [review]
There are a few `"Y"` upper case in the `kernel_build_config_ocil` macro that should probably be in lower case.

**Comment by jan-cerny** [comment]
@ggbecker Great! Thanks!

**Pull request merged**